### PR TITLE
Fix edge case at end of meta prefix

### PR DESCRIPTION
**Pull request by tamird**
Rebase/rework of #1171 since it seems to be important. Closes #1171.

### CODE REVIEW

File: keys/keys.go [hunk 1]
```diff
@@ -289,5 +289,10 @@ func DecodeRangeMetaKey(key proto.Key) (proto.Key, proto.Key) {
 	}
 	// Otherwise find the first entry greater than the given key in the same meta prefix.
 	metaPrefix := proto.Key(key[:len(Meta1Prefix)])
+	metaPrefixMax := proto.MakeKey(metaPrefix, proto.KeyMax)
+
+	if key.Equal(metaPrefixMax) {
+		return key, metaPrefix.PrefixEnd()
+	}
```
Comment (petermattis) [inline keys/keys.go]: `DecodeRangeMetaKey` and this particular case deserves a test.

**Comment by petermattis** [comment]
Any reason this is so urgent we can't wait for #1171 to be merged? Is it affecting some of the open issues? If yes, they should be referenced here.

**Comment by tamird** [comment]
There was a mention of #1193 being related, but I'm not sure that's the case anymore.

File: keys/keys.go [hunk 2]
```diff
@@ -289,5 +289,10 @@ func DecodeRangeMetaKey(key proto.Key) (proto.Key, proto.Key) {
 	}
 	// Otherwise find the first entry greater than the given key in the same meta prefix.
 	metaPrefix := proto.Key(key[:len(Meta1Prefix)])
+	metaPrefixMax := proto.MakeKey(metaPrefix, proto.KeyMax)
```
Comment (tschottdorf) [inline keys/keys.go]: this is static, so it could just live outside of the function call.

File: storage/range_test.go [hunk 3]
```diff
@@ -2384,32 +2384,39 @@ func TestRangeDanglingMetaIntent(t *testing.T) {
...
 	defer tc.Stop()
-
-	reply := proto.InternalRangeLookupResponse{}
-	if err := tc.store.ExecuteCmd(context.Background(), client.Call{
-		Args: &proto.InternalRangeLookupRequest{
-			RequestHeader: proto.RequestHeader{
-				RaftID: 1,
-				Key:    proto.Key{},
+	for _, key := range []proto.Key{
+		// test that InternalRangeLookup works correctly with the first range
```
Comment (tschottdorf) [inline storage/range_test.go]: s/test/Test/

File: keys/keys.go (hunk 2 above)
Comment (tschottdorf) [inline keys/keys.go]: nevermind. Not static.

File: storage/range_test.go [hunk 4]
```diff
@@ -2384,32 +2384,39 @@ func TestRangeDanglingMetaIntent(t *testing.T) {
...
+		// (StartKey==KeyMin). Normally we look up this range in gossip instead of
+		// executing the RPC, but InternalRangeLookup is still used when up-to-date
+		// information is required.
+		proto.KeyMin,
+		// test that InternalRangeLookup works correctly with the last key in a
```
Comment (tschottdorf) [inline storage/range_test.go]: ditto

**Comment by petermattis** [comment]
I think the comment about #1193 being related was a guess. Pretty sure that this change doesn't fix that issue.

File: keys/keys.go (hunk 1 above)
Comment (tamird) [inline keys/keys.go]: Not sure what you mean - this particular case is added to `TestInternalRangeLookup` in this commit.

Comment (petermattis) [inline keys/keys.go]: `TestInternalRangeLookup` is doing an RPC and a whole bunch of other stuff. I'm suggesting adding `TestDecodeRangeMetaKey` to `keys_test.go`.

**Comment by tschottdorf** [comment]
Just want to mention that now that we're comparing against KeyMax on every range lookup, it would seem harsh to have KeyMax be really long, (#437). @petermattis?

File: keys/keys.go (hunk 1 above)
Comment (tamird) [inline keys/keys.go]: Ah, good call. Done.

**Comment by petermattis** [comment]
The comparison against `KeyMax` isn't a problem if it is really long, but the construction of `metaPrefixMap` would be. Good to keep in mind.

**Comment by tamird** [comment]
So the (comment on `DecodeRangeMetaKey`)[https://github.com/cockroachdb/cockroach/blob/master/keys/keys.go#L284] reads: "...The given key must be a valid RangeMetaKey."

And (the tests)[https://github.com/cockroachdb/cockroach/blob/master/keys/keys_test.go#L132:L133] define both `Meta1KeyMax` and `Meta2KeyMax` as invalid `RangeMetaKey`s. Should we be enforcing this invariant in `DecodeRangeMetaKey`?

**Comment by petermattis** [comment]
@tamird `TestValidateRangeMetaKey` defines `proto.MakeKey(Meta{1,2}Prefix, KeyMax)` as valid. Am I misunderstanding your question?

**Comment by tschottdorf** [comment]
I'm confused now. I thought the way you looked up the range was that you take your key, prepend the meta1 prefix, do a scan (starting right after that key), get from that the descriptor and go to that range with meta2prefix+yourkey, scanning right after. Which means that you're never going to have metaXprefix+KeyMax. Am I wrong?

**Comment by tamird** [comment]
FWIW this change was made in 5e9a9c6260277c42aeedfea14856c7f48da3ca0c

**Comment by petermattis** [comment]
@tschottdorf That sounds right to me. @spencerkimball?

**Comment by spencerkimball** [comment]
Well, you could start with yourkey=KeyMax, which I assumed was what we were safeguarding against here. We don't actually enforce yourkey < KeyMax until the actual request is received at the range.

@tamird where was this problem actually coming up in practice.

**Comment by tschottdorf** [comment]
see https://github.com/cockroachdb/cockroach/pull/1037#issuecomment-106957887

**Comment by tamird** [comment]
Added a fix to `ValidateRangeMetaKey`, I think this is good to go now. PTAL

File: keys/keys.go [hunk 5]
```diff
@@ -260,34 +258,37 @@ func ValidateRangeMetaKey(key proto.Key) error {
...
+			return NewInvalidRangeMetaKeyError("body of meta1 range lookup is > KeyMax", key)
+		}
+		return nil
+	}
+	return NewInvalidRangeMetaKeyError("not a meta key", key)
 }
 
-// DecodeRangeMetaKey returns the bounds within which the desired meta
-// record can be found. The given key must be a valid RangeMetaKey.
+// DecodeRangeMetaKey returns the bounds within which the desired meta record
```
Comment (tschottdorf) [inline keys/keys.go]: "MetaScanBounds" is a better name?

Comment (tschottdorf) [inline keys/keys.go]: s/bounds/key range/

File: keys/keys.go [hunk 6]
```diff
@@ -260,34 +258,37 @@ func ValidateRangeMetaKey(key proto.Key) error {
...
+// can be found. The given key must be a valid RangeMetaKey as defined by
```
Comment (tschottdorf) [inline keys/keys.go]: s/found/found by means of an engine scan/

File: keys/keys_test.go [hunk 7]
```diff
@@ -119,6 +119,44 @@ func TestRangeMetaKey(t *testing.T) {
 	}
 }
 
+func TestMetaPrefixLen(t *testing.T) {
+	// DecodeRangeMetaKey and ValidateRangeMetaKey depend on the invariant that
```
Comment (tschottdorf) [inline keys/keys_test.go]: Move that comment up above the test and
```
// TestMetaPrefixLen asserts that both levels of meta keys have the same prefix length,
// as DecodeRangeMetaKey and ValidateRangeMetaKey depend on this fact.
```

File: storage/range_test.go [hunk 8]
```diff
@@ -2384,32 +2384,38 @@ func TestRangeDanglingMetaIntent(t *testing.T) {
 	}
 }
 
-// TestInternalLookupFirstRange tests that InternalRangeLookup works correctly
-// with the first range (StartKey==KeyMin). Normally we look up this range
-// in gossip instead of executing the RPC, but InternalRangeLookup is still
-// used when up-to-date information is required.
-func TestInternalRangeLookupFirstRange(t *testing.T) {
+// TestInternalLookup tests that InternalRangeLookup works correctly.
```
Comment (tschottdorf) [inline storage/range_test.go]: s/TestInternalLookup/TestInternalRangeLookup/ but you might just lose the comment if that's all you wanna say.

**Comment by tschottdorf** [comment]
LGTM. You can just merge after you've addressed any subset of my comments.

**Pull request merged**